File: 프로그래머스/2/87946. 피로도/피로도.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <iostream>

using namespace std;

vector<bool> v;
vector<vector<int>> d;
int ans = 0, sz = 0;

void dfs(int tk, int cnt){
    bool flag = 1;
    for(int i = 0; i < sz; i++) {
        if(!v[i]) {
            flag = 0;
            break;
        }
    }
    if(flag){
        if(cnt > ans) ans = cnt;
        return;
    }
    
    for(int i = 0; i < sz; i++){
        if(v[i]) continue;
        v[i] = 1;
        if(tk >= d[i][0]) dfs(tk - d[i][1], cnt + 1);
        else dfs(tk, cnt);
        v[i] = 0;
    }
}

int solution(int k, vector<vector<int>> dungeons) {
    d = dungeons;
    sz = dungeons.size();
    v.resize(sz, 0);
    
    dfs(k, 0);
    
    return ans;
}

File: 프로그래머스/2/87946. 피로도/피로도.java
import java.util.*;

class Solution {
    boolean[] visit = new boolean[9];
    int answer = -1;
    
    public int solution(int k, int[][] dungeons) {
        Arrays.fill(visit, false);
        
        for(int i = 0; i < dungeons.length; i++){
            if(k < dungeons[i][0]) continue;
            visit[i] = true;
            bt(k - dungeons[i][1], 1, dungeons);
            visit[i] = false;
        }
        
        return answer;
    }
    
    public void bt(int pirodo, int cnt, int[][] dungeons){
        if(cnt > answer) answer = cnt;
        
        for(int i = 0; i < dungeons.length; i++){
            if(pirodo < dungeons[i][0] || visit[i]) continue;
            visit[i] = true;
            bt(pirodo - dungeons[i][1], cnt + 1, dungeons);
            visit[i] = false;
        }
    }
}